acq_stat_reports/utils.py

```diff
@@ -0,0 +1,538 @@
+        The number of realizations of the series to use for the Monte Carlo estimation.
+
+    Returns
+    -------
+    array-like
+        The quantiles
+    """
+    if quantiles is None:
+        quantiles = [15.9, 50, 84.1]
+    samples = (
+        np.random.uniform(size=n_realizations * len(p_acq_model)).reshape(
+            (-1, len(p_acq_model))
+        )
+        < p_acq_model[None]
+    )
+    n = np.sum(samples, axis=1)
+    return np.percentile(n, quantiles)
+
+
+def get_histogram_quantile_ranges(  # noqa: PLR0915
+    data,
+    bin_edges,
+    extra_cols=(),
+    n_samples=10000,
+    success_column="acqid",
+    prob_column="p_acq_model",
+):
+    """
+    Summarize data in bins and calculate quantiles.
+
+    This function expects a Table with `success_column` and `prob_column` columns. The
+    `success_column` column tells whether this sample is a success, and `prob_column` is the
+    a-priori probability of success.
+
+    The function bins the data according to the columns given in `bin_edges`, and creates the
+    following columns:
+    - `bin`: the unique bin index
+    - `n`: the number of samples in the bin
+    - `{col}_low`: the lower edge of the bin (repeated for all bin_edges keys)
+    - `{col}_high`: the upper edge of the bin (repeated for all bin_edges keys)
+    - `{col}`: the center of the bin (repeated for all bin_edges keys)
+    - `{col}_delta`: the width of the bin (repeated for all bin_edges keys)
+    - `{success_column}`: the number of "good" samples in the bin
+    - `median`: the median
+    - `sigma_1_low`: the lower 1-sigma quantile (15.9%)
+    - `sigma_1_high`: the upper 1-sigma quantile (84.1%)
+    - `sigma_2_low`: the lower 2-sigma quantile (2.27%)
+    - `sigma_2_high`: the upper 2-sigma quantile (97.7%)
+
+    NOTE: the result is NOT sparse (it will contain all bins, even if there are no samples in it).
+
+    Parameters
+    ----------
+    data : Table
+        The data to summarize. Must include a `prob_column` column and the corresponding columns
+        for the bin edges.
+    bin_edges : dict
+        A dictionary of bin edges for each column.
+    n_samples : int, optional
+        The number of realizations to use in the Monte Carlo estimation of the expected successes.
+    success_column: str, optional
+        The name of the column containing the success/failure boolean flag.
+    prob_column: str, optional
+        The name of the column containing the probabilities.
+
+    Returns
+    -------
+    Table
+        The quantiles of the data in the bins.
+    """
+    cols = list(bin_edges)
+    extra_cols = list(extra_cols)
+    bin_cols = [f"{bin_col}_bin" for bin_col in cols]
+    for bin_col in cols:
+        bin_edges[bin_col] = np.atleast_1d(bin_edges[bin_col])
+        if len(bin_edges[bin_col].shape) != 1:
+            raise ValueError("bin_edges must be a dict of 1D arrays")
+
+    # data = data.copy()
+    bin_edges = bin_edges.copy()
```

Comment:
This is the part related to my answer above (the apparent misunderstanding).

This does an n-dimensional binning. The code is hard to understand, and it was at this point when I decided I really needed to add a test. I think it can be improved, but it will not necessarily be "simpler".

Reply:
> It looks like this application only uses 1-d binning (am I missing something?). That shaped my understanding of what this function should be doing and the docs did not change my prior.

I have made 2d histograms, to make plots as a function if T_ccd and magnitude. These plots have not been good enough, so they never made it to the final reports. I don't intend to remove this functionality.

The code is not easy to read. I agree. It might be better using an off-the-shelf function, be it `binned_statistic_dd` or something else. I don't think I would use `binned_statistic_dd` though.

Again: I do not intend to remove the n-dimensional binning functionality. We did not write formal requirements before I wrote it, but it is one of the requirements in my view.

Reply:
Great, I'm looking forward to seeing an updated version that uses OTS code for the heavy lifting, hopefully from a widely-used package already in Ska.